### DISCUSSION TITLE
I can't get good qulity of adversarial examples by DeepFool.

### DISCUSSION THREAD

**Discussion by Jiang010** [Trusted-AI/adversarial-robustness-toolbox#1946] [Q&A]
These parameters  are setted for DeepFool. 
umm, I had also tried others parameters. But, the result is not good.
![image](https://user-images.githubusercontent.com/101046093/204843934-ccc1e088-043d-43a8-a3f8-9c5a716c9dbb.png)

The left figure is normal example. And the righ figure is adversarial example, it looks so bad.
![image](https://user-images.githubusercontent.com/101046093/204843650-7e5d2874-38f9-4933-a7b7-6d62f44e0d9f.png)

please, who can tell me, how to improve this question?

**Comment by beat-buesser** [comment]
Hi @Jiang010 Does your model/classifier predict probabilities or logits?

**Reply by Jiang010** [reply]
Hi @beat-buesser thx you respond my question.
this is my classifier code.
```python
    Conv2D(filters=36, kernel_size=7, activation='relu', input_shape= im_shape, padding="same"),
    MaxPooling2D(pool_size=2),
    Conv2D(filters=54, kernel_size=5, activation='relu'),
    MaxPooling2D(pool_size=2),
    Flatten(),
    .
    .
    .
    Dense(20, activation='softmax')  
```

I use the "softmax" as activation. I think that my model output is probabilities.
By the way, I had also tried the "sigmoid" as activation. The DeepFool can work.
It should be caused by this reason.

So, DeepFool can't run in my project.

**Reply by beat-buesser** [reply]
Hi @Jiang010 Can you try with `activation=linear`? I think this attack works best with logits output.

**Comment by Jiang010** [comment]
hi, @beat-buesser, 
Yeah, just like you say. Using the activation="linear" can well run the DeepFool.   
if the activation="linear" not be used, the data of specific label can't well work. It will get terrible results...

umm.. Using the DeepFool may occur some troble in my project.


activation="linear"
```python
cnn_model= Sequential([
     Conv2D(filters=36, kernel_size=7, activation='linear', input_shape= im_shape, padding="same"),
     MaxPooling2D(pool_size=2),
     Conv2D(filters=54, kernel_size=5, activation='linear'),
     MaxPooling2D(pool_size=2),
     Flatten(), 
     Dense(1024, activation='linear'),
     Dense(512, activation='linear'),
     Dense(20, activation='softmax')  
     ])
```
![image](https://user-images.githubusercontent.com/101046093/205964402-39f5d816-24c6-4a6c-807a-11d0174f1595.png)
![image](https://user-images.githubusercontent.com/101046093/205960084-77465180-fa10-4f2e-8054-3f3e2608754b.png)

activation="sigmoid"
``` python 
cnn_model= Sequential([
    Conv2D(filters=36, kernel_size=7, 
                activation='relu', 
                input_shape= im_shape, 
                padding="same"),
    MaxPooling2D(pool_size=2),
    Conv2D(filters=54, 
                kernel_size=5, 
                activation='relu'),
    MaxPooling2D(pool_size=2),
    Flatten(),
    Dense(1024, activation='relu'),
    Dense(512, activation='relu'),
    Dense(20, activation='sigmoid')  
])
```
![image](https://user-images.githubusercontent.com/101046093/205964081-0f2d8c9a-49b5-43da-b361-429493cf32cb.png)
![image](https://user-images.githubusercontent.com/101046093/205959778-d390f2dc-a993-4e7a-8808-7989ab0b1dc8.png)

**Reply by beat-buesser** [reply]
That's great! I think the reason is the algorithm of DeepFool is expecting logits and not probabilities. The implementation of DeepFool algorithm can run with probabilities, but it will be much less successful.
